### ISSUE TITLE
Support pubsub for Actors

### ISSUE THREAD

**Issue by amanbha** [dapr/dapr#501]
## Describe the proposal
Support pubsub between actors, currently the publisher can be a service or an actor. But when it comes to subscribing, the subscribe is supported at service level. We need to extend this to support subscribe at an Actor level. An actor should be able to subscribe to a message of type "Message_One" with its actor id and when that message is published a specific method is invoked on that particular actor id. If multiple actor ids have subscribed for that message then method will be invoked for all the subscribed actor ids.

**Comment by mchmarny** [comment]
Given the unique need for some kind of mapping between inbound event and actor (based on the event type and and some kind of logic to derive ID to create/activate actor), I wonder if this possibly better handled in new kind of binding that can be configured with ref to pubsub binding name as a "source" in metadata. Trying to handle this in PubSub construct could potentially add complexity to the simple pubsub construct we support now.

**Comment by amanbha** [comment]
@mchmarny  The implementation for ACtors pubsub will built on top of Dapr's pubsub so the description is incorrect. I will modify it keep and this issue for bindings only.
@LMWF Could you open another issue for Actors pubsub?

**Comment by amanbha** [comment]
@mchmarny @LMWF With the proposal for Actors pubsub, it will not require logic to determine id from the event. The Actors layer in dapr runtime will save that information and will build on top of Dapr's pubsub building block.

**Comment by yaron2** [comment]
> @mchmarny @LMWF With the proposal for Actors pubsub, it will not require logic to determine id from the event. The Actors layer in dapr runtime will save that information and will build on top of Dapr's pubsub building block.

Yeah, but this proposal assumes the actor can first subscribe to events only after it gets activated via service invocation.

@mchmarny is talking about the scenario where event come in (either through bindings or Pub/Sub) that need to trigger an actor irrespective of it being created earlier.

**Comment by amanbha** [comment]
Yes, these are 2 scenarios.

**Comment by LMWF** [comment]
Note #1601 above was forked for bindings.

**Comment by artursouza** [comment]
I think there is only one scenario: subscription for ActorType+ActorId. The other scenarios are a specific scenario of this one. The reason why is that Dapr does not track which actors exist since they are virtual actors. The fact that an actor is active does not mean it only exists in that period of time. Virtual actors "always" exist.

So, there must be a call to say "subscribe these actors Ids of this actor type to this topic". There is no way to say "subscribe all actors of this type to this topic" because there are infinite number of possible instances of an actor type. The same reason why there is no way to "subscribe all actors of all types to this topic" - there are infinite number of those.

ActorType is a limited and well defined list provided by the app. Subscribing an ActorId for all actorTypes to a topic seems too specific and can be implemented by simply repeating the subscriptions of that ActorId for each of the known Actor Types. 

Long story short, Dapr needs the ActorType and ActorId to process a subscription. So, maybe the Cases 1, 2 and 3 can be redefined  as:

1. "subscribe and ActorType + ActorId to a particular topic"
2. "subscribe an ActorType to a particular topic where the Actor Id is defined by an attribute in the message."
3. Repeat Case 2 above for each known actor type + use pubsub filtering if needed.

**Comment by yaron2** [comment]
Another thing to think about here is how an actor expresses the [pub/sub response](https://docs.dapr.io/reference/api/pubsub_api/#expected-http-response) for acking/nacking/retrying the request. This might need a change in the programming model that'll make it's way to the SDKs, assuming we want to support it.

Another approach would be to use a simplified event model for actors pub/sub where a successful actor method invocation is considered an ACK and an error returned is considered NACK.

**Comment by pkedy** [comment]
Given the nature of how most pubsub brokers work, a single topic subscription can be shared by several clients (e.g. Kafka topic partitions balanced across consumers). This means some component will need to (re)route the message to the node that is serves (or will serve) that specific Actor Type/ID. If this component invokes the actor via the current architecture, then it is responsible for ACK/NACK and possibly DLQ as well.

This “pattern” is referred to as an Arbiter if you want to use the terminology from the inventor of Actors.
https://www.youtube.com/watch?v=7erJ1DV_Tlo

**Comment by lloydfischer** [comment]
If ordering is critical then there cannot be multiple clients. Actors provide a nice way to ensure that messages are handled in order. Imagine enrollment changes for a family - they must be processed in order. Routing them to an actor for the family solves that but only if they arrive at the actor in order. A single consumer for the partition can guarantee that. If there are multiple consumers then the time will come when messages arrive at the actor out of order.

**Comment by pkedy** [comment]
@lloydfischer Correct, that is something to be very mindful of.

To add on, speaking strictly in the case of Kafka, the key of the message usually determines which partition it is written to. A single partition is assigned to a single consumer that processes the messages one-by-one and in order. In other words, the topic partitions are the unit of parallelism in Kafka. What I meant above is that there can be multiple topic partitions being processed, and those nodes (1 per partition) likely are not serving the target actor and thus use service invocation to deliver the payload. And to @lloydfischer's point, the partitioning scheme needs to be considered to ensure in-order processing.

I suppose there could be an affinity in that the consuming node could become the node assigned to that actor, but that seems like it would limit the scale of actors.

**Comment by yaron2** [comment]
The orchestration should happen at the `daprd` level. individual components should not know to send messages to nodes, pods, or if they are serving actors or not.

The pub/sub events will be delivered using actor service invocation and will obey the single threaded nature of actors.
Ordering of messages is decoupled from the actor functionality. If we decide ordered delivery needs to be enabled, we can tweak `daprd` to start a single consumer. This consumer does not have to live on the same pod or even the same node as the actor instance.

**Comment by lloydfischer** [comment]
Completely agree that guaranteeing order is best done in the runtime. And
to echo your earlier comment, that means that the ack can only go back to
the transport if we are certain that the message will be delivered, in
order, to the actor. I worry about the wait in the channel, for instance.
If the runtime crashes, are the messages in the channel lost? We can
recover them by replay but guaranteeing order is hard since there may be
others in process on their way to the channel.  We may need persistence in
the channel to protect against that.   The more distance from transport to
actor the harder it is to ensure the order is maintained properly. Any step
along the way where we can be forced to replay is a place where order is at
risk.

Obviously, flipping the flag to guarantee order means accepting a
performance hit, but when you need it, it will be worth it. To me, this is
the killer app for actors.

**Comment by pkedy** [comment]
To clarify, by "some component" I mean some piece of code running in `daprd` that is responsible for actor service invocation and optionally ensuring in-order processing (but I think in-order processing depends on the behavior of the underlying broker). So I think we are on the same page there.

For handling runtime crashes (or other cases where messages are sent again such as rebalancing), my thought is that in the actor would implement idempotency such that previously processed messages are safely ignored. Currently, all pubsubs in Dapr are durable so you would just need to pick one that best supports in-order processing.

**Comment by lloydfischer** [comment]
Agreed that idempotency is a separate thing. For daprd the key, as Yaron noted, is that there be a single consumer. More generally there must be always exactly one path for a message to reach an actor. If there is a possibility for a message to take a sneak path it will find it. Any secondary path will be have the potential for a message to pass another and arrive out of order. Multiple arrivals are fine provided the first arrival of any message is before the first arrival of a message later in the message ordering. 
I don't pretend to understand all the plumbing but it could be that restricting to a single consumer may be enough to guarantee a single path. 
As to idempotency, and I'm thinking out loud here, I wonder if a partition actor could play this role. This actor would have the id of the partition so all messages from the partition would arrive there. It can keep state to remember the messages it has seen and drop the duplicates. Then, in my use case, that partition actor can delegate to the family actor for further processing. 
If we set the "ordered" flag then the runtime would ensure only one path to the partition actor. Even if we don't preserve order I think the family actor is enjoying "exactly once" semantics.

**Comment by halspang** [comment]
Bringing up a design here that I've discussed with a few people but would like to open up to a broader audience.

The goal of this proposal is to determine a simple-case path forward that let's Dapr provide value while also keeping the door open for future enhancements.

## Design
### High Level Choices
There are several key elements that should be called out before the actual design:
1. This feature is an extension of Actors (Event driven invocation), not Pubsub.
1. The Actor's host will provide the subscription configuration.
1. The Actor can be activated from a pubsub invocation.
1. Events should be pushed to the Actors.
1. Scalability of Actor should be maintained.

This design does not attempt to solve issues like mass invocation from a single message or topic/message routing. Rather it intends to be open enough that it can easily adopt into planned routing solutions and more complex orchestration in the future.

### Sequence Diagram
![actor_pubsub](https://user-images.githubusercontent.com/70976921/129283818-0249ec86-99bd-49c0-b58e-70450341de43.png)

### Publishing to an Actor
Publishing to an Actor has to include several pieces of metadata including the Actor Type, Actor Id, and method to be invoked by the subscriber. As this feature is event-driven actor invocation, we'll expose a new API that mimics the existing method invocation API.

```
POST http://localhost:<daprPort>/v1.0/actors/<actorType>/<actorId>/publish/<pubsubname>/<topic>/<method>
```

By including all of the actor information in the request, the resulting CloudEvent can have the `ActorId` encoded into it. Items like the `ActorType` and `Method` may not need to be encoded as they can be matched to a component/topic on the subscription side before the method is invoked.

#### Inferred ActorId
An potential enhancement (or potentially part of the core offering) could be allowing the user to specify a field that already exists in the CloudEvent to be used as an `ActorId`. An example of this could be an IoT scenario. It is already known that the IoT realm is a strong proponent of Actors and event-driven programming. In many cases, the messages could include fields like `DeviceId` which could then be mapped directly to an Actor. Instead of having them specify this directly, we could provide this functionality directly.

**Note:** I do not think this scenario is mutually exclusive from the direct Actor invocation stated above. I believe they can exist in tandem, the question here is whether this should be part of the initial offering or an incremental change as we move to support the broader set of Pubsub use-cases.

### Subscription Configuration
Actors will provide a type/id to subscription configuration during their startup process. This will piggy-back onto the existing actor configuration which Dapr loads via a call to `dapr/config` during app startup. By doing this here, we can guarantee that the actor that is subscribing to the system is actively being hosted. Should that app/dapr instance die it becomes much easier to cleanup the resources.

This does create two distinct areas where pubsub subscriptions are managed. Though non-ideal, it is better than the alternate (discussed below). It is also important to reframe this past Actors with Pubsub and see it more as an event/trigger configuration for Actors. We already know of at least one more trigger scenario with [bindings for actors](https://github.com/dapr/dapr/issues/1601).

The alternate option here would be to specify this entirely in the subscription CRD. This was not chosen as it decouples the subscriptions with the actor creation/registration. At the time of reading the subscription CRD, there is no way to know what actors are supported unless they are already up. Keeping a subscription loop up consistently while waiting is wasteful and trying to restart the subscription when the actor does get registered creates more points where subscriptions start and makes the system less consistent.

#### Example
```json
{
  "entities":["actorType1", "actorType2"],
  "actorIdleTimeout": "1h",
  "actorScanInterval": "30s",
  "drainOngoingCallTimeout": "30s",
  "drainRebalancedActors": true,
  "pubsub": [
    {
      "name": "myActorPubsub",
      "topic": "myPubsubTopic",
      "actorType": "actorType1",
      "method": "pubsubHandler"
    }
  ]
}
```
**Note:** The method of specifying the topic/message route may change based on the broader Pubsub filtering design.

### Actor Invocation
Actor invocation will happen very similarly to a general pubsub subscription today. When a subscription is received, Dapr begins a goroutine that occasionally reads from the given topic and issues the supplied callback when a message is encountered. 

The goal here is to keep as much logic in the Dapr runtime as possible and keep it in-line with current system style. When a new Actor subscription is received, it will mimic the existing behavior, simply altering the callback to invoke the Actor instead of the service endpoint.

Since Actors are single-threaded and can be long running, we will most likely need to fire off these events as new goroutines which could increase our load. As the Actor is locked though, we do not need to worry about messages being read in parallel unless sent to different Ids. If, on the other hand, pubsub is already setup to have long running subscription reads, we may not have to worry about this. 

ACKing/NACKing responses would, at first, have to be a bit simpler than the current implementations. A successful invocation would be considered an ACK and an invocation that returns in error would be considered a NACK. Overtime, we can standardize a means for Actors to better communicate their status back to the Dapr runtime.

/cc @artursouza @yaron2

**Comment by theperm** [comment]
Good work, 
is the example what the alternate CRD option would have been? That is not clear.

'Actors will provide a type/id to subscription configuration during their startup process.'
Do you have an idea of how or where this would take place. Can you clarify if it would be encapsulated in the Actor Class/Type (in C# anyway) and read one time when an Actor class is registered? Assuming this is before and doesn't require an Actor to be activated. First message could activate an actor for first time?

Just a side note there is on open feature for on-demand subscriptions - https://github.com/dapr/dapr/issues/814
so worth keeping in mind we don't want to limit Actors to a one time static setup if dynamic subscriptions are supported. I imagine we can add this later on top of the static subscriptions.

**Comment by yaron2** [comment]
That's a very good proposal @halspang.

I want to make sure that we allow actors to be invoked via pub/sub in two ways:

1. By explicitly publishing messages to `POST http://localhost:<daprPort>/v1.0/actors/<actorType>/<actorId>/publish/<pubsubname>/<topic>/<method>
`.
2. By receiving events from any general purpose pub/sub component

Can you confirm this design supports actors being invoked with pub/sub messages that do not originate from the new API above?

**Comment by halspang** [comment]
> Good work,
> is the example what the alternate CRD option would have been? That is not clear.
>

The example is actually the actor configuration.

> 'Actors will provide a type/id to subscription configuration during their startup process.'
> Do you have an idea of how or where this would take place. Can you clarify if it would be encapsulated in the Actor Class/Type (in C# anyway) and read one time when an Actor class is registered? Assuming this is before and doesn't require an Actor to be activated. First message could activate an actor for first time?

That's actually a leftover from the original design and we've now moved to not including the Id in the configuration. The type will still be included similarly to how it is today. To denote a pubsub invocation, this is what I'm thinking for the .NET sdk.

```csharp
/// <summary>
/// Interface for Actor invocation.
/// </summary>
public interface IDemoActor : IActor
{
    Task Say();

    [Topic("pubsub", ”handler")]
    Task DoPubsub(string person);
}

// Send event via ActorProxy.
var proxy = ActorProxy.Create<IDemoActor>(actorId, "DemoActor");
await proxy.DoPubsub("Someone");
```

> Just a side note there is on open feature for on-demand subscriptions - #814
> so worth keeping in mind we don't want to limit Actors to a one time static setup if dynamic subscriptions are supported. I imagine we can add this later on top of the static subscriptions.

This will be interesting to work with and will require some adaptation when it happens. I'll keep on eye on it, but at least for now this will stand as an actor's feature and we can incorporate other methodologies as they come.

**Comment by halspang** [comment]
> That's a very good proposal @halspang.
> 
> I want to make sure that we allow actors to be invoked via pub/sub in two ways:
> 
> 1. By explicitly publishing messages to `POST http://localhost:<daprPort>/v1.0/actors/<actorType>/<actorId>/publish/<pubsubname>/<topic>/<method> `
> 2. By receiving events from any general purpose pub/sub component provided there's a CloudEvent envelope associated with the message
> 
> Can you confirm this design supports actors being invoked with pub/sub messages that do not originate from the new API above?
> 
> I'm looking at the `Inferred Actor ID` part and it looks to cover that scenario. I'd like to see a mechanism that allows users to specify which field in a CloudEvent message corresponds to an Actor ID, and ideally have this in the first implementation phase.
> 
> This will allow users to leverage actors with _existing systems_ much easier without needing to rewrite large parts of their code. In addition, in some systems the pub/sub component is used to consume events from publishers that are not owned by the team of developers that own the subscribers.

Yes, that's the field that we're proposing to handle this case. I put it there to see if there's a strong desire for it which you've answered. I think we can get it into the initial version of the feature, though ideally we'd implement them separately but in the same milestone, just to reduce PR size/blast radius.

**Comment by halspang** [comment]
After presenting at the community call, one topic that came up was broadcasting to multiple actors from a single message. Dapr has a couple of options for orchestration:

- Allow for `ActorId/inferredActorId` to be an array and call each actor accordingly.
- Have an event for each actor.

Another ask was to see how low touch we could make the initial feature in the SDKs. Can this just be a configuration and the actors will start taking events?

**Comment by georgestevens99** [comment]
Hal, Thanks for the good presentation!

I would like to chime in be sure you plan on having some performance/load tests available as you develop, and thereafter as well.  

Why?  Because actors are often used by the dozens, and even the hundreds or thousands in some cases.  And having pubsub capabilities can well act to amplify the load costs of actors on all or parts of a system.  For example it would be completely "normal" to expect to publish an event subscribed to by 50 actors, while it would be quite unusual to expect publish an event subscribed to by 50 apps.  An example of such an event would be "StartCollectingData" published after system initialization.  

So, to make sure potential users of this feature approach it with realistic expectations, the results of some load/performance tests would be highly useful.  And, as this feature is being developed such tests will also help in several ways, one of the main ones being discovering bottlenecks that can potentially be fixed before feature delivery.

Thanks.

**Comment by jjcollinge** [comment]
@halspang just a quick comment after watching the community call, I noticed you intended to add the attributes `dapr.actorType` and `dapr.actorId` to the CloudEvent payload. In order to keep the CloudEvent aligned with the spec we'd need to use the [CloudEvent attribute naming conventions](https://github.com/cloudevents/spec/blob/v1.0.1/spec.md#context-attributes) which are lowercase and only allow characters from [a-z0-9]. That would mean using something like `dapractorid` and `dapractortype` instead.

**Comment by aktxyz** [comment]
+100 for this feature !!  it sounds like something was presented or demo'd in a community call, any recollection which month?

**Comment by yaron2** [comment]
> +100 for this feature !! it sounds like something was presented or demo'd in a community call, any recollection which month?

https://youtu.be/5TTZV-HB10Y

**Comment by theperm** [comment]
I just got round to watching the community call and reading this reply.
> ```cs
> /// <summary>
> /// Interface for Actor invocation.
> /// </summary>
> public interface IDemoActor : IActor
> {
>     Task Say();
> 
>     [Topic("pubsub", ”handler")]
>     Task DoPubsub(string person);
> }
> 
> // Send event via ActorProxy.
> var proxy = ActorProxy.Create<IDemoActor>(actorId, "DemoActor");
> await proxy.DoPubsub("Someone");
> ```

This approach has sender explicitly call the pubsub handler for an actor via proxy. The advantage here is you can call an Actor asynchronously, as was stated it creates an inbox for an actor. That is a good thing.
Because its explicit the producer knows where they will be sending, this is generally an event driven antipattern which is why i think the inferred ActorID will be the more important method. A service could be outside of Dapr or just want to publish a message via the ordinary pubsub component. So how would we have an actor subscribe to those messages? 

Adding the DeviceId or ActorID to the CloudEvent is one approach.
But that means the producer would have to ensure it is added to the CloudEvent.
Typically I've seen the destination ActorId inferred from topics, like Users/UserA
Is that also a viable approach? It might better handle messages from outside Dapr. I guess the topic is on the CloudEvent source right? 
for example in C#SDK we could interpolate the subscription declaratively, and then infer the actorid when the pattern matchs
```cs
Topic("User/{ActorId}", ”handler")] 
or 
[Topic("*/{ActorId}", ”handler", )] if wild cards where supported
```
I think this would also allow different actor types to subscribe to the same topic as they would not be competing consumers.

**Comment by halspang** [comment]
@theperm - Thanks for the feedback, answers are below!

> Because its explicit the producer knows where they will be sending, this is generally an event driven antipattern which is why i think the inferred ActorID will be the more important method. A service could be outside of Dapr or just want to publish a message via the ordinary pubsub component. So how would we have an actor subscribe to those messages?

As you state, we plan on having the ability to infer the actor id via the actor's configuration. This is the same configurations actors already provide with the added `pubsub` object. 
```json
{
  "entities":["actorType1", "actorType2"],
  "actorIdleTimeout": "1h",
  "actorScanInterval": "30s",
  "drainOngoingCallTimeout": "30s",
  "drainRebalancedActors": true,
  "pubsub": [
    {
      "name": "myActorPubsub",
      "topic": "myPubsubTopic",
      "actorType": "actorType1",
      "method": "pubsubHandler"
      "inferredActorId": "deviceId"
    }
  ]
}
```

This does require that the producer know to put the field into the message. Ideally, the field will already be there if we're working with an existing system that someone wishes to map to an event based actor approach. 

> Typically I've seen the destination ActorId inferred from topics, like Users/UserA. Is that also a viable approach?

This is a valid scenario, but not one that I think we can support yet. Similar to standard pubsub topics, Dapr starts listening to a given topic at startup. This have to be strongly defined by the listening application during startup. As such, we don't currently support dynamic subscriptions which seem like they are required here. Once that is supported, we could definitely start looking into it.

It's also worth noting that the `Topic` attribute in the Actor space is used to reference where to publish, not where to subscribe. In this scenario it seems like you want the existing usage of `Topic` and would then have the controller call the actor. This is technically something you could do today, but again you'd need the dynamic topics. Am I understanding how you want to use that correctly?

**Comment by theperm** [comment]
Great, that makes sense. I didn't originally pick up that the topic was for the new publish mechanism.

> It's also worth noting that the `Topic` attribute in the Actor space is used to reference where to publish, not where to subscribe. In this scenario it seems like you want the existing usage of `Topic` and would then have the controller call the actor. This is technically something you could do today, but again you'd need the dynamic topics. Am I understanding how you want to use that correctly?

Yes you understand correctly. And as to your suggestion of what is technically possible today (skipping dynamic topics for now). I've built a .Net web controller (in the same process as the actor host) that subscribe to topics and then relay those to the actors by extracting ActorID from the message. The challenge with this approach is when you scale the process, the received messages (ingress from pubsub) are load balanced across all instances of the controller. This creates the potential for messages (at higher throughputs) to get out of order before they get relayed to an actor. So the responsibility and complexity of guaranteeing message ordering falls on the app developer. On top of that the actor will create back pressure in the controller, which defeats one advantage of using messaging in the first place.. The hope is by making this pattern native to the actor block that it can handle message ordering as one less concern for the developer. 

Have you gave any thought how to scale message processing and guarantee ordering? Do you want to continue on that topic now?

**Comment by theperm** [comment]
Keep open bot.

**Comment by aktxyz** [comment]
agree with keep open bot

**Comment by ryou90** [comment]
Too bad that this feature is constantly postponed, although it is so highly rated. I've been waiting for it for so long...

**Comment by yaron2** [comment]
> Too bad that this feature is constantly postponed, although it is so highly rated. I've been waiting for it for so long...

@artursouza @halspang what's the reason for moving this to 1.7?

**Comment by aktxyz** [comment]
ugg bummer this got bumped again

**Comment by artursouza** [comment]
> > Too bad that this feature is constantly postponed, although it is so highly rated. I've been waiting for it for so long...
> 
> 
> 
> @artursouza @halspang what's the reason for moving this to 1.7?

Given the priorities in 1.6, we simply do not have the resources to deliver this feature in 1.6. We can move the date for 1.6 but we are already overshooting. The amount of work to deliver resiliency and more stable components + missing features in SDKs is already taking all our capacity for 1.6. This was a hard cut. I am happy to help someone from the community to deliver this in 1.6 if we have volunteers.

**Comment by yaron2** [comment]
> > > Too bad that this feature is constantly postponed, although it is so highly rated. I've been waiting for it for so long...
> > 
> > 
> > @artursouza @halspang what's the reason for moving this to 1.7?
> 
> Given the priorities in 1.6, we simply do not have the resources to deliver this feature in 1.6. We can move the date for 1.6 but we are already overshooting. The amount of work to deliver resiliency and more stable components + missing features in SDKs is already taking all our capacity for 1.6. This was a hard cut. I am happy to help someone from the community to deliver this in 1.6 if we have volunteers.

@artursouza I triaged to 1.6 and assigned myself. This is best effort, if I don't get to it we'll move this to 1.7.

**Comment by halspang** [comment]
@yaron2 - I have a PoC locally. I can post it up if you want a starting point. It needs some clean-up and testing but is otherwise pretty close to ready.

**Comment by et1975** [comment]
Could someone elaborate on what ideal back pressure story is like in dapr?
I'm also curious how processing guarantees in all-awaitable chain of Fabric actors vs fire-n-forget models of execution are to be maintained.

**Comment by theperm** [comment]
Currently every call to an actor in a chain requires every call to return
before the next can proceed. So this causes back pressure on every actor.
Even tho the calls are asynchronous (non blocking) you are effectively
making a synchronous call to the last actor in the chain.

There is no fire and forget if you want introduce asynchronous calls say in
the middle of the chain so the end actors don't exert back pressure on the
preceding actors.

This is the feature that is required. Pub sub support would enable it or
alternatively introduce inbox queues to actors that can execute
asynchronously.

You can do it today by introducing a pub sub "controller" in the middle but
it's extra engineering that we shouldn't need to do and it's not without
its challenges either.

Hope that helps.

**Comment by et1975** [comment]
Thanks, but I understand how things works when everything is based on RPC transport. What I'm asking is how this proposal will maintain the back pressure and processing guarantees if the transport is based on unbounded queues.
What I'm getting at is that making the transports simply a matter of configuration may seem attractive, but might be a mistake - they are not interchangeable as far as abstractions that rely on them go.

**Comment by jason-visotrust** [comment]
Why would it need backpressure? The actor is going to pull messages at its own rate?

**Comment by yaron2** [comment]
Dapr will consume the pub/sub message and then invoke the actor as if it were a synchronous RPC message, meaning it will respect the turn based concurrency of the actor.

**Comment by et1975** [comment]
To be clear my concerns stem from the "send", not "consume" side of things.

> Dapr will consume the pub/sub message and then invoke the actor as if it were a synchronous RPC message, meaning it will respect the turn based concurrency of the actor.

The concern is fault handling, here's the scenario: Actor A -queue-> Actor B.

From A's perspective, the invocation to B is complete and successful the moment dapr publishes it to the broker. This changes the semantics of the interaction from trustworthy (B's done it's work) to untrustworthy (A has no idea if B will ever do the work, let alone complete it successfully). 

> Why would it need backpressure? The actor is going to pull messages at its own rate?

Indeed, the dapr->actor interaction would implement backpressure by the virtue of RPC interaction between them. But the same is not true in actor->dapr interaction - as a source of messages such actor has no idea what happens downstream. 

A competing consumer on a queue could keep the optimal load on the processors, but an actor that simply floods downstream will pick the work w/o or not its downstream is ready to process or has been failing for a while and we're now in a data loss situation. 

Between these two concerns a sender cannot trust that its message was processed w/o rebuilding RPC (ack/nack/timeout) on top of messaging, which could be suboptimal from many angles. 

What I'm proposing is that "send" is not implemented as means of invoking another actor over messaging.

**Comment by ryou90** [comment]
Removed from 1.6. but not in Milestone 1.7?

**Comment by georgestevens99** [comment]
I have been watching this feature out of the corner of my eye for months now.  It is very popular, but looks as if it has gotten stalled out for the last couple releases.  I hope work to complete this can resume after 1.7 is out the door and result in getting something nicely useable (but maybe not with every feature including the kitchen sink) released and into developers hands in the next few months.

**Comment by gormal** [comment]
Any update on this? It's not part of Milestone 1.8?

**Comment by DanielHazlewood** [comment]
Would also like to chime in, this feature would be amazing.

**Comment by berndverst** [comment]
We have an intern working on this over the summer. Most likely it will not be included in the 1.9 release, but there is a good chance it will be out in 1.10.

**Comment by yaron2** [comment]
> Would also like to chime in, this feature would be amazing.

There's a PR open and work is in progress but this will not make it to 1.9.

**Comment by ryou90** [comment]
Hope to finally see this feature in 1.10 Milestone?

**Comment by Andemki** [comment]
@berndverst @yaron2 Are there any updates on this item?

**Comment by onionhammer** [comment]
Updates?

**Comment by msislam3** [comment]
Any updates on the release? It would be a great feature to have for us as we need some way to asynchronously send messages between actors

**Comment by theperm** [comment]
I agree. Instead of adding new features to dapr it would be better to
improve the existing ones and make them fit for purpose.

Lack of inbox or pubsub for an actor is severely limiting its use case.

**Comment by yaron2** [comment]
We recognize this is a highly requested feature and will do our best to enable it. In order to ensure this and other features related to actors behave optimally and deliver the best possible experience to users, it was crucial to improve actors all up which serve as the basis for higher level APIs today such as workflows. Many improvements went into actors in the last 3 releases and it might be high time to re-examine pub/sub for actors now.

**Comment by yaron2** [comment]
> Also reinvesting in the .NET sdk for actors would be good; for a relatively new library DataContractSerializer is a weird choice for the serializer
> 
> I hate this <img alt="image" width="782" src="https://user-images.githubusercontent.com/969938/261631556-dc9affb4-1740-4774-91bb-df175181aa8c.png">

cc @halspang

**Comment by olitomlinson** [comment]
My 2c :  This work should be broken up into multiple phases

- **Phase 1** Solve async Actors at-scale (message passing, rather than blocking calls) - this will make _many_ adopters happy.
- **Phase 2** Only once async is solved, look at what you can build on top of it, i.e. PubSub/FanOut/Routing etc

**Comment by theperm** [comment]
> My 2c : This work should be broken up into multiple phases
> 
> * **Phase 1** Solve async Actors at-scale (message passing, rather than blocking calls) and provide a first-class DevEx in the SDKs for fire & forget messaging - this will make _many_ adopters happy.
> * **Phase 2** Only once async is solved, look at what you can build on top of it, i.e. PubSub/FanOut/Routing etc

What he said!

**Comment by joshuadmatthews** [comment]
Just want to +1 this request. A proper Saga implementation should be based on an actor that handles to messages imo.

**Comment by yaron2** [comment]
> Just want to +1 this request. A proper choreographed saga implementation should be based on an actor that handles messages imo. Durable entity azure functions provide essentially this, but I’d love to stay Dapr only.

For Saga, have you checked Dapr Workflows? Might be a better higher level programming model for that

**Comment by joshuadmatthews** [comment]
@yaron2 Yes I have been closely watching workflows, I wasn’t sure Saga was the best fit though as it isn’t one of the documented patterns discussed.

**Comment by olitomlinson** [comment]
@joshuadmatthews 

> I guess looking back the External system interaction example is basically describing a saga.

Pretty much!

FWIW I had a play with making a very basic [saga-style wrapper](https://github.com/olitomlinson/dapr-workflow-examples/blob/kafka2/Workflow/Saga.cs) for Dapr Workflows, which [worked out kinda neat](https://github.com/olitomlinson/dapr-workflow-examples/blob/kafka2/Workflow/Workflows/SagaWorkflow.cs)

**Comment by georgestevens99** [comment]
Nice use of Generics, Oli!  That looks great.
